internal/overlord/logstate/loki.go

```diff
@@ -0,0 +1,204 @@
+	"time"
+
+	"github.com/canonical/pebble/internal/logger"
+	"github.com/canonical/pebble/internal/plan"
+	"github.com/canonical/pebble/internal/servicelog"
+)
+
+const lokiParserSize = 4 * 1024
+
+type lokiForwarder struct {
+	serviceName string
+	targetName  string
+	target      plan.LogTarget
+
+	iterator servicelog.Iterator
+	parser   *servicelog.Parser
+	stopCh   chan struct{}
+
+	buffer *lokiBuffer
+	client *lokiClient
+}
+
+func newLokiForwarder(
+	serviceName, targetName string, target plan.LogTarget, iterator servicelog.Iterator,
+) (*lokiForwarder, error) {
+
+	labels := map[string]string{
+		"service": serviceName,
+		"manager": "pebble",
+	}
+
+	return &lokiForwarder{
+		serviceName: serviceName,
+		targetName:  targetName,
+		target:      target,
+		iterator:    iterator,
+		parser:      servicelog.NewParser(iterator, lokiParserSize),
+		buffer:      newLokiBuffer(labels, 10, 1*time.Second),
+		client:      newLokiClient(target.Location),
+	}, nil
+}
+
+func (f *lokiForwarder) Forward() {
+	for {
+		// Check if Stop has been called
+		select {
+		case <-f.stopCh:
+			f.flushBuffer()
+			return
+		default:
+			// continue
+		}
+
+		if f.buffer.ReadyToFlush() {
+			f.flushBuffer()
+		}
+
+		if f.parser.Next() {
+			f.buffer.WriteEntry(f.parser.Entry())
+		} else {
+			if err := f.parser.Err(); err != nil {
+				logger.Noticef(
+					"Forwarding logs to target %q: cannot parse log for service %q: %v",
+					f.targetName, f.serviceName, err)
+			}
+		}
+	}
+}
+
+func (f *lokiForwarder) Stop() {
+	f.stopCh <- struct{}{}
+}
+
+func (f *lokiForwarder) flushBuffer() {
+	err := f.client.Send(f.buffer.Request())
+	if err != nil {
+		logger.Noticef(
+			"Forwarding logs to target %q: cannot parse log for service %q: %v",
+			f.targetName, f.serviceName, err)
+		return
+	}
+
+	f.buffer.Empty()
+}
+
+type lokiBuffer struct {
+	labels      map[string]string
+	messages    []string
+	maxMessages int
+
+	deadline *time.Time
+	delay    time.Duration
+}
+
+func newLokiBuffer(labels map[string]string, maxMessages int, delay time.Duration) *lokiBuffer {
+	return &lokiBuffer{
+		labels:      labels,
+		messages:    make([]string, 0, maxMessages),
+		maxMessages: maxMessages,
+		deadline:    nil,
+		delay:       delay,
+	}
+}
+
+func (b *lokiBuffer) WriteEntry(entry servicelog.Entry) {
+	// Write in format
+	//    [ "<unix epoch in nanoseconds>", "<log line>" ]
+	timestamp := fmt.Sprint(entry.Time.UnixNano())
+	message := strings.Trim(entry.Message, "\n")
+	encodedMsg := fmt.Sprintf("[%q,%q]", timestamp, message)
+	b.messages = append(b.messages, encodedMsg)
+
+	if b.deadline == nil {
+		// Set write deadline
+		deadline := time.Now().Add(b.delay)
+		b.deadline = &deadline
+	}
+}
+
+func (b *lokiBuffer) ReadyToFlush() bool {
+	return (len(b.messages) >= b.maxMessages) ||
+		(b.deadline != nil && time.Now().After(*b.deadline))
+}
+
+// Prepares a JSON request for sending to Loki.
+func (b *lokiBuffer) Request() io.Reader {
+	labelStrings := []string{}
```

Comment:
This could be optimised in the future.

Reply:
In what way?

Reply:
All the `fmt.Sprintf` and `strings.Join` and appends and whatnot are doing allocations and string formatting, so we want to avoid that as much as possible when it's a fairly tight loop (once per log, or once per flush).

This would be done by instantiating a `bytes.Buffer` once per `lokiBuffer` (but reset on flush), writing directly to that buffer, which would avoid string allocations. Also, `fmt.Sprintf` is quite slow as it has to parse the format string, allocate, and so on, so let's avoid that.

Pseudo-code:

```go
var buf bytes.Buffer  # initialize this once

# write this prefix (including labels) at the start, and after each flush
buf.Write([]byte(`{"streams":[{"stream":{"service":"foo"},"values":[`))

# write the ["timestamp","message"] part each log line
buf.Write(jsonEncodedLog)

# write this suffix before each flush
buf.Write([]byte(`]}]}`))

# to actually write to request, pass &buf directly to client.Post
```

Reply:
ok have done this, I've used `buf.WriteString(str)` instead of `buf.Write([]byte(str))`, not sure if it makes a difference.